osu.Server.Spectator/Hubs/EntityStore.cs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Extensions.ObjectExtensions;
+
+namespace osu.Server.Spectator.Hubs
+{
+    /// <summary>
+    /// Tracks and ensures consistency of a collection of entities that have a related permanent ID.
+    /// </summary>
+    /// <typeparam name="T">The type of the entity being tracked.</typeparam>
+    public class EntityStore<T>
+        where T : class
+    {
+        private readonly Dictionary<long, TrackedEntity> entityMapping = new Dictionary<long, TrackedEntity>();
+
+        private const int lock_timeout = 5000;
+
+        /// <summary>
+        /// Retrieve an entity with a lock for use.
+        /// </summary>
+        /// <param name="id"></param>
+        /// <param name="createOnMissing"></param>
+        /// <returns>An <see cref="ItemUsage{T}"/> which allows reading or writing the item. This should be disposed </returns>
+        /// <exception cref="KeyNotFoundException">Thrown if <see cref="createOnMissing"/> was false and the item is not in a tracked state.</exception>
```

Comment:
Hmm... this xmldoc is not 100% true as written. Even with `createOnMissing = true` a consumer can get this exception by accidentally winning the race on the `entityMapping` lock (gets entity from store before it's cleaned itself up). For 100% safety in that case there should just be a retry.

Reply:
Should be resolved now.